File: src/runtime/boxes/bb_breakx.c
/* _XBRKX    BREAKX      like BRK but fails on zero advance */
#include "bb_box.h"
#include <stdlib.h>
#include <string.h>


spec_t bb_breakx(void *zeta, int entry)
{
    brkx_t *ζ = zeta;
    spec_t BREAKX;
    if (entry==α)                                                               goto BREAKX_α;
    if (entry==β)                                                               goto BREAKX_β;
    BREAKX_α:       ζ->δ=0;                                                     
                    while (Δ+ζ->δ<Ω && !strchr(ζ->chars,Σ[Δ+ζ->δ])) ζ->δ++;     
                    if (ζ->δ==0 || Δ+ζ->δ>=Ω)                                   goto BREAKX_ω;
                    BREAKX = spec(Σ+Δ,ζ->δ); Δ += ζ->δ;                         goto BREAKX_γ;
    BREAKX_β:       Δ -= ζ->δ;                                                  goto BREAKX_ω;
    BREAKX_γ:                                                                   return BREAKX;
    BREAKX_ω:                                                                   return spec_empty;
}

brkx_t *bb_breakx_new(const char *chars)
{ brkx_t *ζ=calloc(1,sizeof(brkx_t)); ζ->chars=chars; return ζ; }

File: src/runtime/boxes/BbBreakx.java
package snobol4.runtime.boxes;

/**
 * BbBreakx.java — BREAKX: like BRK but fails on zero advance
 * Port of bb_breakx.c / bb_breakx.s
 *
 *   BREAKX_α:  δ=0; while (Δ+δ<Ω && !chars.has(Σ[Δ+δ])) δ++;
 *              if (δ==0 || Δ+δ>=Ω)             goto BREAKX_ω;
 *              BREAKX=spec(Σ+Δ,δ); Δ+=δ;       goto BREAKX_γ;
 *   BREAKX_β:  Δ-=δ;                            goto BREAKX_ω;
 */
public class BbBreakx extends BbBox {
    private final String chars;
    private int          delta;

    public BbBreakx(MatchState ms, String chars) { super(ms); this.chars=chars; }

    @Override public Spec alpha() {
        delta = 0;
        while (ms.delta + delta < ms.omega &&
               chars.indexOf(ms.sigma.charAt(ms.delta + delta)) < 0) delta++;
        if (delta == 0 || ms.delta + delta >= ms.omega) return null;
        Spec r = new Spec(ms.delta, delta);
        ms.delta += delta;
        return r;
    }

    @Override public Spec beta() { ms.delta -= delta; return null; }
}
